lib/Common/Memory/RecyclerPointers.h

```diff
@@ -216,21 +214,24 @@ class WriteBarrierPtr
 
     // Getters
     T * operator->() const { return ptr; }
-    operator T*() const { return ptr; }
+    operator T* const & () const { return ptr; }
 
-    T const** AddressOf() const { return &ptr; }
+    T* const * AddressOf() const { return &ptr; }
     T** AddressOf() { return &ptr; }
 
-    T const** operator&() const
-    {
-        static_assert(false, "Might need to set barrier for this operation, and use AddressOf instead.");
-        return &ptr;
-    }
-    T** operator&()
+    // Taking immutable address is ok
```

Comment:
Talked offline. @leirocks's idea is to examine and baseline our usage of const_cast to prevent future incorrect use related to this code. That is doable. Will add a note here.